### PR TITLE
precount outstanding requests so that request and status are sized pr…

### PR DESCRIPTION
**Pull request by jedwards4b**
…operly

### CODE REVIEW

File: src/clib/pio_darray.c [hunk 1]
```diff
@@ -1357,32 +1359,37 @@ int flush_output_buffer(file_desc_t *file, bool force, PIO_Offset addsize)
...
+      reqcnt+=vdesc->nreqs;
+      if(vdesc->nreqs>0) maxreq = i;
+    }
+    int request[reqcnt];
+    int status[reqcnt];
 
-
+    for(int i=0; i<=maxreq; i++){
+      vdesc = file->varlist+i;
 #ifdef MPIO_ONESIDED
```
Comment (edhartnett) [inline src/clib/pio_darray.c]: What is MPIO_ONESIDED?

File: src/clib/pio_darray.c [hunk 2]
```diff
@@ -1357,32 +1359,37 @@ int flush_output_buffer(file_desc_t *file, bool force, PIO_Offset addsize)
...
+    int reqcnt;
+    maxreq = 0;
+    reqcnt=0;
+    rcnt=0;
     for(int i=0; i<PIO_MAX_VARS; i++){
       vdesc = file->varlist+i;
+      reqcnt+=vdesc->nreqs;
+      if(vdesc->nreqs>0) maxreq = i;
+    }
+    int request[reqcnt];
```
Comment (edhartnett) [inline src/clib/pio_darray.c]: Wow I didn't even know you could do that in C. I would have thought that a malloc was required. Didn't know you could declare an array with a size you just computed!

File: src/clib/pio_darray.c (hunk 1 above)
Comment (jedwards4b) [inline src/clib/pio_darray.c]: Early this year I worked on IO performance with ROMIO (MPI-IO) and PnetCDF developers.
The MPIO_ONESIDED optimization went into the ROMIO library - that optimization should eventually make it into the mpich distribution but as far as I know that hasn't happened yet.  
It improves performance but introduces restrictions on what you can put into the write buffer - 
the buffer (aggregated across IO tasks) must represent a contiguous block on the file.  This ifdef is about trying to assure that that is the case.

File: src/clib/pio_darray.c (hunk 2 above)
Comment (jedwards4b) [inline src/clib/pio_darray.c]: It's part of the c99 standard.

File: src/clib/pio_darray.c (hunk 1 above)
Comment (edhartnett) [inline src/clib/pio_darray.c]: OK, so I should make this settable from cmake, as I did for PIO_USE_MALLOC?

I will take care of it in the next merge...

**Pull request merged**